pkg/target/http.go

```diff
@@ -215,63 +337,150 @@ func AdaptHTTPTargetFunc(f func(c *HTTPTargetConfig) (*HTTPTarget, error)) HTTPT
 
-	for _, msg := range safeMessages {
-		request, err := http.NewRequest("POST", ht.httpURL, bytes.NewBuffer(msg.Data))
-		if err != nil {
-			errResult = multierror.Append(errResult, errors.Wrap(err, "Error creating request"))
-			failed = append(failed, msg)
-			continue
-		}
-		request.Header.Add("Content-Type", ht.contentType)                // Add content type
-		addHeadersToRequest(request, ht.headers, ht.retrieveHeaders(msg)) // Add headers if there are any
-		if ht.basicAuthUsername != "" && ht.basicAuthPassword != "" {     // Add basic auth if set
-			request.SetBasicAuth(ht.basicAuthUsername, ht.basicAuthPassword)
-		}
-		requestStarted := time.Now().UTC()
-		resp, err := ht.client.Do(request) // Make request
-		requestFinished := time.Now().UTC()
+	for _, chunk := range chunks {
+		grouped := ht.groupByDynamicHeaders(chunk)
 
-		msg.TimeRequestStarted = requestStarted
-		msg.TimeRequestFinished = requestFinished
+		for _, group := range grouped {
+			var reqBody []byte
+			var goodMsgs []*models.Message
+			var badMsgs []*models.Message
+			var err error
 
-		if err != nil {
-			errResult = multierror.Append(errResult, err)
-			failed = append(failed, msg)
-			continue
-		}
-		defer resp.Body.Close()
-		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
-			sent = append(sent, msg)
-			if msg.AckFunc != nil { // Ack successful messages
-				msg.AckFunc()
+			if ht.requestTemplate != nil {
+				reqBody, goodMsgs, badMsgs = ht.renderBatchUsingTemplate(group)
+			} else {
+				reqBody, goodMsgs, badMsgs = ht.provideRequestBody(group)
 			}
-		} else {
-			errResult = multierror.Append(errResult, errors.New("Got response status: "+resp.Status))
-			failed = append(failed, msg)
-			continue
+
+			invalid = append(invalid, badMsgs...)
+
+			if len(goodMsgs) == 0 {
+				continue
+			}
+
+			request, err := http.NewRequest("POST", ht.httpURL, bytes.NewBuffer(reqBody))
+
+			if err != nil {
+				panic(err)
+			}
+
+			request.Header.Add("Content-Type", ht.contentType)                        // Add content type
+			addHeadersToRequest(request, ht.headers, ht.retrieveHeaders(goodMsgs[0])) // Add headers if there are any - because they're grouped by header, we just need to pick the header from one message
+			if ht.basicAuthUsername != "" && ht.basicAuthPassword != "" {             // Add basic auth if set
+				request.SetBasicAuth(ht.basicAuthUsername, ht.basicAuthPassword)
+			}
+			requestStarted := time.Now()
+			resp, err := ht.client.Do(request) // Make request
+			requestFinished := time.Now()
+
+			// Add request times to every message
+			for _, msg := range goodMsgs {
+				msg.TimeRequestStarted = requestStarted
+				msg.TimeRequestFinished = requestFinished
+			}
+
+			if err != nil {
+				failed = append(failed, goodMsgs...)
+				errResult = multierror.Append(errResult, errors.New("Error sending request: "+err.Error()))
+				continue
+			}
+			defer resp.Body.Close()
+
+			if resp.StatusCode >= 200 && resp.StatusCode < 300 {
```

Comment:
Assumption check - the new functionality for configurable retries can only be used for a non-200 status code because of this line.

I don't think that's incorrect, just confirming that I understand correctly to make sure it gets documented accurately.

---

Just as food for thought - I wonder if we might consider different behaviour in future. The Braze API can have scenarios where there's a 2XX response, but the data doesn't actually get ingested. At present I think of this as problematic behaviour from the API, so I don't think it's wise to introduce complexity to handle it before we consider and exhaust other ways to deal with that scenario. Just calling out this possible future nuance in case ideas about it bubble up in the meantime.

Reply:
I think it's the right call, this comment was more food for thought about the future than anything :)